cmd/main.go

```diff
@@ -2,16 +2,188 @@ package main
+	klog.InitFlags(klogFlags)
+	if err := klogFlags.Parse([]string{fmt.Sprintf("-stderrthreshold=%d", klogLevel), "-v=2", "-logtostderr=false"}); err != nil {
+		klog.Fatal(err.Error())
+	}
+	snapshotURL := getEnvWithDefault("AES_SNAPSHOT_URL", fmt.Sprintf(DefaultSnapshotURLFmt, ExternalSnapshotPort))
+	diagnosticsURL := getEnvWithDefault("AES_DIAGNOSTICS_URL", fmt.Sprintf(DefaultDiagnosticsURLFmt, AdminDiagnosticsPort))
+
+	reportDiagnostics := os.Getenv("AES_REPORT_DIAGNOSTICS_TO_CLOUD")
+	if reportDiagnostics == "true" {
+		ambAgent.SetReportDiagnosticsAllowed(true)
+	}
+
+	metricsListener, err := net.Listen("tcp", ":8080")
 	if err != nil {
-		fmt.Fprintf(os.Stdout, "error executing from argparser: %v", err)
-		os.Exit(1)
+		klog.Fatal(err.Error())
+	}
+	dlog.Info(ctx, "metrics service listening on :8080")
+
+	grp := dgroup.NewGroup(ctx, dgroup.GroupConfig{})
+	grp.Go("metrics-server", func(ctx context.Context) error {
+		metricsServer := agent.NewMetricsServer(ambAgent.MetricsRelayHandler)
+		return metricsServer.Serve(ctx, metricsListener)
+	})
+
+	// use a Go context so we can tell the leaderelection code when we
+	// want to step down
+	leaseCtx, leaseCancel := context.WithCancel(ctx)
+	defer leaseCancel()
+
+	// listen for interrupts or the Linux SIGTERM signal and cancel
+	// our context, which the leader election code will observe and
+	// step down
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-ch
+		klog.Info("Received termination, signaling shutdown")
+		leaseCancel()
+	}()
+
+	// each call to the leaselock should have a unique id
+	id := uuid.New().String()
+
+	// we use the Lease lock type since edits to Leases are less common
+	// and fewer objects in the cluster watch "all Leases".
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      leaseLockName,
+			Namespace: agentNamespace,
+		},
+		Client: clientset.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+	}
+
+	// use a go context to kill watchers OnStoppedLeading
+	var (
+		watchCtx    context.Context
+		watchCancel context.CancelFunc
+		i           int
+	)
+	run := func() {
+		i += 1
+		grp.Go(fmt.Sprintf("watch-%v", i), func(ctx context.Context) error {
+			watchCtx, watchCancel = context.WithCancel(ctx)
+			return ambAgent.Watch(watchCtx, snapshotURL, diagnosticsURL)
+		})
+	}
+
+	// start the leader election code loop
+	leaderelection.RunOrDie(leaseCtx, leaderelection.LeaderElectionConfig{
+		Lock: lock,
+		// IMPORTANT: you MUST ensure that any code you have that
+		// is protected by the lease must terminate **before**
+		// you call cancel. Otherwise, you could have a background
+		// loop still running and another process could
+		// get elected before your background loop finished, violating
+		// the stated goal of the lease.
+		ReleaseOnCancel: true,
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   15 * time.Second,
```

Comment:
Should probably set to 45 seconds; from Luke:
```...Set RenewDeadline to 2/3 of LeaseDuration, and RetryPeriod to 2/15 of  LeaseDuration (those are the ratios used in kubernetes.git).```

Reply:
aha. interesting.